### PR TITLE
Limit size of PagesIndex

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

**Review by sopel39 — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/operator/PagesIndex.java [hunk 1]
```diff
@@ -218,6 +220,9 @@ public void addPage(Page page)
 
         for (int position = 0; position < page.getPositionCount(); position++) {
             long sliceAddress = encodeSyntheticAddress(pageIndex, position);
+            if (valueAddresses.size() >= 2_000_000_000) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/PagesIndex.java:223]: Should we use https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#MIN_VALUE instead?

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/operator/PagesIndex.java:223]: @sopel39  `valueAddresses.size()` is an `int`, so surely kess than Integer.MAX_VALUE

@electrum where does 2B come from? safety measure?
please add a code comment

Comment (electrum) [reply presto-main/src/main/java/io/prestosql/operator/PagesIndex.java:223]: I was picking a round number, like we do in other places (search for "1 billion")

Comment (electrum) [reply presto-main/src/main/java/io/prestosql/operator/PagesIndex.java:225]: Added a comment

**Comment by gauravkm** [comment]
Thanks for the fix! We might be hitting this (currently on Release 338). We saw instances of OOM's due to `Array size exceeds VM Limit`. (More details here https://prestosql.slack.com/archives/CGB0QHWSW/p1607393717357700?thread_ts=1607028047.311900&cid=CGB0QHWSW)

I am curious as to whether this is a recent regression? We were earlier on Release 322 and rarely saw OOM's.

**Comment by gauravkm** [comment]
@electrum Could you provide more context on :

1. What the PagesIndex does?
2. Why it should never grow beyond this size?
3. Would this cause only the problematic query to fail or more widespread failures? 

Not sure, If there is a write up or issue detailing why this change was made. Would be very useful for us to understand a bit more about this.

**Comment by electrum** [comment]
1. `PagesIndex` is a data structure that contains the address of every value position in a list of pages. It used for hash tables for joins, sorting, etc.
2. For performance reasons, a single Java array is used for the list of positions. Java arrays are fundamentally limited to `Integer.MAX_VALUE` items, since they are indexed by signed integers. The actual limit is JVM dependent and is often smaller, due to object overhead, but `MAX_VALUE-8` is a safe limit. Unfortunately, the JVM treats an attempted allocation of an array over the limit as `OutOfMemoryError`, so it's important to never allocate one too large.
3. With this fix, it will cause the offending query to terminate gracefully, rather than crash the JVM.

If you are hitting this error, it is likely because you have a join with more than two billion items on a single machine. The solution is to add more machines so that the items are spread otu.

**Pull request merged**